src/main/java/com/woowacamp/soolsool/core/member/service/MemberService.java

```diff
@@ -21,26 +30,49 @@ public class MemberService {
 
     private final MemberRepository memberRepository;
     private final MemberRoleRepository memberRoleRepository;
+    private final MemberMileageChargeRepository memberMileageChargeRepository;
 
     @Transactional
     public void addMember(final MemberAddRequest memberAddRequest) {
-        final MemberRole memberRole = memberRoleRepository.findById(1L)
-            .orElseThrow(() -> new SoolSoolException(MemberErrorCode.MEMBER_NO_ROLE_TYPE));
+        checkDuplicatedEmail(memberAddRequest.getEmail());
+
+        final MemberRole memberRole = getMemberRole(memberAddRequest);
         final Member member = memberAddRequest.toMember(memberRole);
         memberRepository.save(member);
     }
 
+    private void checkDuplicatedEmail(final String email) {
+        final Optional<Member> duplicatedEmil = memberRepository
+            .findByEmail(new MemberEmail(email));
+
+        if (duplicatedEmil.isPresent()) {
+            throw new SoolSoolException(MemberErrorCode.MEMBER_DUPLICATED_EMAIL);
+        }
+    }
+
+    private MemberRole getMemberRole(final MemberAddRequest memberAddRequest) {
+        final MemberRoleType memberRoleType = stream(MemberRoleType.values())
+            .filter(type -> Objects.equals(type.getType(), memberAddRequest.getMemberRoleType()))
+            .findFirst()
+            .orElse(MemberRoleType.CUSTOMER);
+
+        return memberRoleRepository.findByName(memberRoleType)
+            .orElseThrow(() -> new SoolSoolException(MemberErrorCode.MEMBER_NO_ROLE_TYPE));
+    }
```

Comment:
```suggestion
    private MemberRole getMemberRole(final MemberAddRequest memberAddRequest) {
        final MemberRoleType memberRoleType = MemberRoleType
            .findByType(memberAddRequest.getMemberRoleType())
            .orElse(() -> new SoolSoolException(MemberErrorCode.MEMBER_NO_ROLE_TYPE));

        return memberRoleRepository.findByName(memberRoleType)
            .orElseThrow(() -> new SoolSoolException(MemberErrorCode.MEMBER_NO_ROLE_TYPE));
    }
```

코드 가독성을 개선해보고 싶었는데 `valueOf()`가 `IllegalArgumentException`을 던지는 걸 해결하다보니 비슷해졌네요. 한 번 같이 고민해봐요.

또, 구매자-판매자가 아닌 제 3자의 역할로 회원가입할 경우 아예 회원 가입이 안되는건 어떨까요?

Reply:
1. enum 과 db 를 둘다 체킹하는거라서 메서드를 나눌까도 고민을 많이 했던 부분인데 나도 명확한 해결점을 찾진 못했어
2. 우리가 회원 가입 시, default를 "구매자"로 지정했기 때문에 굳이 회원가입을 취소하는 것 보다는 우선 "구매자" 로서 회원 가입 시키는 정책으로 가져가면 어떨까?? 굳이 memberType 때문에 회원가입을 못하는 불상사는 없어도 좋을것 같어^^